fuel-relayer/src/relayer.rs

```diff
@@ -233,40 +185,23 @@ impl Relayer {
                 .to_block(best_block)
                 .address(ValueOrArray::Array(contracts.clone()));
 
-            let logs = self.stop_handle(|| provider.get_logs(&filter)).await??;
-
-            for eth_event in logs {
-                let fuel_event = EthEventLog::try_from(&eth_event);
-                if let Err(err) = fuel_event {
-                    // not formated event from contract
-                    error!(target:"relayer", "Eth Event not formated properly in inital sync:{}",err);
-                    // just skip it for now.
-                    continue;
-                }
-                let fuel_event = fuel_event.unwrap();
-                self.pending
-                    .append_eth_events(&fuel_event, eth_event.block_number.unwrap().as_u64())
-                    .await;
-            }
+            let logs = handle_interrupt!(self, provider.get_logs(&filter))??;
+            self.queue.append_eth_logs(logs).await;
 
             // 3. Start listening to eth events
             let eth_log_filter = Filter::new().address(ValueOrArray::Array(contracts.clone()));
             watchers = Some((
-                self.stop_handle(|| provider.watch_blocks()).await??,
-                self.stop_handle(|| provider.watch(&eth_log_filter))
-                    .await??,
+                handle_interrupt!(self, provider.watch_blocks())??,
+                handle_interrupt!(self, provider.watch(&eth_log_filter))??,
```

Comment:
why recreate the watchers in a loop and wait 500ms each time? 

1. The watchers do nothing unless polled as a stream.
2. The default poll interval is 7000ms.

Reply:
`watch_blocks` and `watch` are sending the request to eth client to subscribe to a particular event here: https://github.com/gakonst/ethers-rs/blob/598c00e55ddbe98267c82de477b355f7f73010b7/ethers-providers/src/provider.rs#L668-L693 
Here is how pub/sub API works: https://geth.ethereum.org/docs/rpc/pubsub
There is a possibility that in-between times I ask for the newest block and start watching events new block gets mined and I can miss it, that is why I get best_block before subscribing and after, and compare if they are the same.
50ms is really arbitrary as to introduce some delay between subscription next request for the best block.

Reply:
When we register the subscriber on the eth client, the eth client will start to queue pending logs and they will become available on the next watcher pull request.

```
notifications are stored in an internal buffer and sent from this buffer to the client. If the client is unable to keep up and the number of buffered notifications reaches a limit (currently 10k) the connection is closed. Keep in mind that subscribing to some events can cause a flood of notifications, e.g. listening for all logs/blocks when the node starts to synchronize.
```

Reply:
I see, so is the idea that we want to reset the subscription buffer on the eth node to clear any log events that could be preemptively discarded when a reorg happens, instead of relying on the reorg handling logic in the finalization queue later on?

This seems like it could congest the eth node since the watchers don't automatically call `uninstall_filter` for their subscription id when dropped. Would this cause each abandoned watcher queue to pile up to 10k events before being cleaned up by the eth node?

Reply:
I'm concerned about relying so heavily on the state of the subscription mechanism and trying to finesse the perfect order of events it returns by all this special case handling around starting and stopping watchers etc. 

For example, if the eth node crashed, restarted, updated, or our sticky session needed to be dropped due to infura rescaling, the relayer would just crap out and stop running. See my other comments about coming up with a more durable solution that avoids the complexities of reorg handling etc.

Reply:
As time goes filter will be droped by eth client.

Here is point where rechecking blocks is needed to be done: https://github.com/FuelLabs/fuel-core/blob/feac06c7f883cbba91df6df1dc05520f878a4ebc/fuel-relayer/src/finalization_queue.rs#L259
Or even before this loop.

For recovery scenearions you can put tasks in the backlog, when they become problem we can focus on them.

Reply:
Wrote up my summary of how I think this should be handled here: https://github.com/FuelLabs/fuel-core/issues/398

I'm good with deferring this as follow-up work, given that we identified use-cases for pending events in the UX.